sdk/program/src/lib.rs

```diff
@@ -1,3 +1,650 @@
+//! implementation to the environment. The `cfg` predicate used for identifying
+//! compilation for on-chain programs is `target_arch = "bpf"`, as in this
+//! example from the `solana-program` codebase that logs a message via a
+//! syscall when run on-chain, and via a library call when offchain:
+//!
+//! [rbpf]: https://github.com/solana-labs/rbpf
+//! [eBPF]: https://ebpf.io/
+//! [cc]: https://doc.rust-lang.org/reference/conditional-compilation.html
+//!
+//! ```
+//! pub fn sol_log(message: &str) {
+//!     #[cfg(target_arch = "bpf")]
+//!     unsafe {
+//!         sol_log_(message.as_ptr(), message.len() as u64);
+//!     }
+//!
+//!     #[cfg(not(target_arch = "bpf"))]
+//!     program_stubs::sol_log(message);
+//! }
+//! # mod program_stubs {
+//! #     pub(crate) fn sol_log(message: &str) { }
+//! # }
+//! ```
+//!
+//! This `cfg` pattern is suitable as well for user code that needs to work both
+//! on-chain and off-chain.
+//!
+//! `solana-program` and `solana-sdk` were previously a single crate. Because of
+//! this history, and because of the dual-usage of `solana-program` for two
+//! different environments, it contains some features that are not available to
+//! on-chain programs at compile-time. It also contains some on-chain features
+//! that will fail in off-chain scenarios at runtime. This distinction is not
+//! well-reflected in the documentation.
+//!
+//! For a more complete description of Solana's implementation of eBPF and its
+//! limitations, see the main Solana documentation on [on-chain programs][ocp].
+//!
+//! [ocp]: https://docs.solana.com/developing/on-chain-programs/overview
+//!
+//! # Core data types
+//!
+//! - [`Pubkey`] - An [ed25519] public key. All Solana accounts are identified
+//!   by a `Pubkey`, though some may not have corresponding private keys. As
+//!   Solana programs do not handle secret keys, the full [`Keypair`] is not
+//!   defined in `solana-program` but in `solana-sdk`.
+//! - [`Hash`] - A [SHA-256] hash, used to uniquely identify blocks.
+//! - [`AccountInfo`] - A description of a single Solana account. All accounts
+//!   that might be accessed by a program invocation are provided to the program
+//!   entrypoint as `AccountInfo`.
+//! - [`Instruction`] - A directive telling the runtime to execute a program,
+//!   passing it a set of accounts and program-specific data.
+//! - [`ProgramError`] and [`ProgramResult`] - The error type that all programs
+//!   must return, reported to the runtime as a `u64`.
+//!
+//! [`Pubkey`]: pubkey::Pubkey
+//! [`Hash`]: hash::Hash
+//! [`Instruction`]: instruction::Instruction
+//! [`AccountInfo`]: account_info::AccountInfo
+//! [`ProgramError`]: program_error::ProgramError
+//! [`ProgramResult`]: entrypoint::ProgramResult
+//!
+//! [ed25519]: https://ed25519.cr.yp.to/
+//! [`Keypair`]: https://docs.rs/solana-sdk/latest/solana_sdk/signer/keypair/struct.Keypair.html
+//! [SHA-256]: https://en.wikipedia.org/wiki/SHA-2
+//!
+//! # Serialization
```

Comment:
It should probably be pointed out that both borsh and bincode are expensive (with the latter being very) and that developers are free to use whatever serialisation or data format they wish for their own account and instruction data.  But also, they for data they did not source (sysvars for example) they need to deserialize with he correct format